[coroutine] Implement llvm.coro.await.suspend intrinsic

fpasserby opened the pull request:
Implement `llvm.coro.await.suspend` intrinsics, to deal with performance regression after prohibiting `.await_suspend` inlining, as suggested in #64945.
Actually, there are three new intrinsics, which directly correspond to each of three forms of `await_suspend`:
```
void llvm.coro.await.suspend.void(ptr %awaiter, ptr %frame, ptr @wrapperFunction)
i1 llvm.coro.await.suspend.bool(ptr %awaiter, ptr %frame, ptr @wrapperFunction)
ptr llvm.coro.await.suspend.handle(ptr %awaiter, ptr %frame, ptr @wrapperFunction)
```
There are three different versions instead of one, because in `bool` case it's result is used for resuming via a branch, and in `coroutine_handle` case exceptions from `await_suspend` are handled in the coroutine, and exceptions from the subsequent `.resume()` are propagated to the caller. 

Await-suspend block is simplified down to intrinsic calls only, for example for symmetric transfer:
```
%id = call token @llvm.coro.save(ptr null)
%handle = call ptr @llvm.coro.await.suspend.handle(ptr %awaiter, ptr %frame, ptr @wrapperFunction)
call void @llvm.coro.resume(%handle)
%result = call i8 @llvm.coro.suspend(token %id, i1 false)
switch i8 %result, ...
```
All await-suspend logic is moved out into a wrapper function, generated for each suspension point.
The signature of the function is `<type> wrapperFunction(ptr %awaiter, ptr %frame)` where `<type>` is one of  `void` `i1` or `ptr`, depending on the return type of `await_suspend`. 
Intrinsic calls are lowered during `CoroSplit` pass, right after the split.

Because I'm new to LLVM, I'm not sure if the helper function generation, calls to them and lowering are implemented in the right way, especially with regard to various metadata and attributes, i. e. for TBAA. All things that seemed questionable are marked with `FIXME` comments.

There is another detail:  in case of symmetric transfer raw pointer to the frame of coroutine, that should be resumed, is returned from the helper function and a direct call to `@llvm.coro.resume` is generated. C++ standard demands, that `.resume()` method is evaluated. Not sure how important is this, because code has been generated in the same way before, sans helper function.

fpasserby committed:
Implement llvm.coro.await.suspend intrinsic

ChuanqiXu9 reviewed (commented):
Thanks for looking into this. I haven't looked it in details. Given this is complex, it should take a relative longer time.

Here is some quick feedbacks:
- Every time we change the non-static data members of AST nodes, we need to update the serializations. Otherwise, modules won't work.
- Every time we change the LLVM intrinsics, we need to change the documentation.
- Every time we change the LLVM with a functional change, we need a test in the LLVM part.

And there is something I haven't fully understand:
- I feel odd about `helperFunction`. I don't understand the necessity and I feel we should make it in the middle end.
- I feel odd about `IsSuspendNoThrow`. Can't we make it simply during the CodeGen time by the attribute of `await_suspend`?.
- I am wondering if we can get rid of the introduction of `OpaqueFramePtr` from `CoroutineSuspendExpr` either in CodeGen part or in Sema part.

But I am not sure if my feelings are correct. I need more time looking into it.

fpasserby committed:
Remove isAwaitSuspendNoThrow

fpasserby committed:
Update AST serializers

fpasserby commented:
Thank you for replying so quickly. 
Your suggestion about `IsSuspendNoThrow` makes sense, it's totally possible to perform the check during codegen. 
About `helperFunction` - this is how I understood the comment in `SemaCoroutine.cpp` near the place where `std::coroutine_handle::address()` is marked as always inline. As far as I can see, it makes perfect sense to hide all the temporary objects till coroutine frame is formed, given the history of optimizations and sanitizers messing it up. For example, current implementation seems to work correctly with the reproducer from #72006.
For other suggestions - I hope, I've fixed up AST serialization and added some documentation. Will add LLVM transformation tests a bit later.

fpasserby committed:
Document added intrinsics

ChuanqiXu9 commented on clang/lib/CodeGen/CGCoroutine.cpp:
```diff
@@ -232,16 +237,59 @@ static LValueOrRValue emitSuspendExpression(CodeGenFunction &CGF, CGCoroData &Co
+            .getScalarVal());
+    SuspendHelperCallArgs.push_back(SuspendHelper);
+
+    auto IID = llvm::Intrinsic::coro_await_suspend;
+    if (S.getSuspendExpr()->getType()->isBooleanType())
+      IID = llvm::Intrinsic::coro_await_suspend_bool;
+    else if (S.getSuspendExpr()->getType()->isVoidPointerType())
+      IID = llvm::Intrinsic::coro_await_suspend_handle;
```
nit: This pattern occurs multiple times. So maybe we can add an enum `SuspendReturnType` to `CoroutineSuspendExpr` and a getter for that. Then we can use a switch here to make it looks better.

fpasserby replied:
Added switch.

ChuanqiXu9 commented on llvm/docs/Coroutines.rst:
```diff
@@ -1744,6 +1744,273 @@ a call to ``llvm.coro.suspend.retcon`` after resuming abnormally.
+  ; after lowering
+  await.suspend:
+    %save = call token @llvm.coro.save(ptr %hdl)
+    ; the call to await_suspend_function is inlined
```
```suggestion
    ; the call to await_suspend_function can be inlined
```

fpasserby replied:
Fixed

ChuanqiXu9 commented on llvm/docs/Coroutines.rst:
```diff
@@ -1744,6 +1744,273 @@ a call to ``llvm.coro.suspend.retcon`` after resuming abnormally.
+The '``llvm.coro.await.suspend``' intrinsic hides C++ `await-suspend`
+block code from optimizations on presplit coroutine body 
+to avoid miscompilations. This version of intrinsic corresponds to 
+'``void awaiter.await_suspend(...)``' variant.
```
Maybe it is better to express the motivation too. It is explained in the comment in SemaCoroutine.

fpasserby replied:
Added motivation (copied that comment almost verbatim)

ChuanqiXu9 commented on llvm/docs/Coroutines.rst:
```diff
@@ -1744,6 +1744,273 @@ a call to ``llvm.coro.suspend.retcon`` after resuming abnormally.
+    %suspend = call i8 @llvm.coro.suspend(token %save, i1 false)
+    ...
+  await.ready:
+    call void @"Awaiter::await_ready"(ptr %awaiter)
```
shouldn't this be `await_resume`? await_ready should happen before await_suspend

fpasserby replied:
Yes, was a typo.

ChuanqiXu9 commented on llvm/docs/Coroutines.rst:
```diff
@@ -1744,6 +1744,273 @@ a call to ``llvm.coro.suspend.retcon`` after resuming abnormally.
+  ; helper function example
+  define void @await_suspend_function(ptr %awaiter, ptr %hdl)
+    entry:
+      %hdl.tmp = alloca %"struct.std::coroutine_handle"
+      %hdl.result.tmp = alloca %"struct.std::coroutine_handle"
+      %hdl.promise.tmp = alloca %"struct.std::coroutine_handle.0"
+      %hdl.promise = call ptr @"std::corouine_handle<promise_type>::from_address"(ptr %hdl)
+      %hdl.promise.tmp.dive = getelementptr inbounds %"struct.std::coroutine_handle.0",
+        ptr %hdl.promise.tmp, i32 0, i32 0
+      %hdl.promise.tmp.dive2 = getelementptr inbounds %"struct.std::coroutine_handle",
+        ptr %hdl.promise.tmp.dive, i32 0, i32 0
+      store ptr %hdl.promise, ptr %hdl.promise.tmp.dive2
+      call void @llvm.memcpy.p0.p0.i64(ptr %hdl.tmp, ptr %hdl.promise.tmp, i64 8, i1 false)
+      %hdl.tmp.dive = getelementptr inbounds %"struct.std::coroutine_handle",
+        ptr %hdl.tmp, i32 0, i32 0
+      %hdl.arg = load ptr, ptr %hdl.tmp.dive
```
The readability looks not nice. Maybe we don't need to paste the complete IR here.

fpasserby replied:
Reduced the amount of IR

ChuanqiXu9 commented on llvm/lib/Transforms/Coroutines/CoroSplit.cpp:
```diff
@@ -79,6 +79,73 @@ using namespace llvm;
 namespace {
 
+// Created on demand if the coro-early pass has work to do.
+class Lowerer : public coro::LowererBase {
```
Why do we need to inherit `coro::LowererBase`? Can't we make it directly in a function?

fpasserby replied:
Yes, it is possible. Switched to a simple function.

ChuanqiXu9 commented on llvm/lib/Transforms/Coroutines/CoroSplit.cpp:
```diff
@@ -79,6 +79,73 @@ using namespace llvm;
+    HelperCall->setAttributes(NewAttributes);
+    HelperCall->setDebugLoc(Call->getDebugLoc());
+    NewCall = HelperCall;
+  }
```
```suggestion
  } else
      llvm_unreachable(...)
```

nit: personal interest

fpasserby replied:
Added unreachable

ChuanqiXu9 commented on llvm/lib/Transforms/Coroutines/CoroSplit.cpp:
```diff
@@ -79,6 +79,73 @@ using namespace llvm;
+  CB->replaceAllUsesWith(NewCall);
+  CB->eraseFromParent();
+
+  InlineFunctionInfo FnInfo;
+  auto InlineRes = InlineFunction(*NewCall, FnInfo);
+  assert(InlineRes.isSuccess() && "Expected inlining to succeed");
+  (void)InlineRes;
```
Don't do this. This should be done by later inlining pass. We prefer one pass do one thing strategy.

fpasserby replied:
Removed manual inlining

ChuanqiXu9 commented on llvm/lib/Transforms/Coroutines/CoroSplit.cpp:
```diff
@@ -79,6 +79,73 @@ using namespace llvm;
+  (void)InlineRes;
+}
+
+void Lowerer::lowerAwaitSuspends(Function &F) {
```
We can collect CoroAwaitSuspendInst during the construction of Shape. So that we can save one iterations of the function.

fpasserby replied:
Moved collection to the construction of Shape

ChuanqiXu9 commented on llvm/docs/Coroutines.rst:
```diff
@@ -1744,6 +1744,273 @@ a call to ``llvm.coro.suspend.retcon`` after resuming abnormally.
+                ptr %awaiter,
+                ptr %hdl,
+                ptr @await_suspend_function)
+    call void @llvm.coro.resume(%next)
```
I feel better if there is `@llvm.coro.suspend`

fpasserby replied:
Added `@llvm.coro.suspend`

ChuanqiXu9 commented on llvm/lib/Transforms/Coroutines/CoroSplit.cpp:
```diff
@@ -1513,6 +1580,11 @@ static void handleNoSuspendCoroutine(coro::Shape &Shape) {
+    // but additional analysis before the check should ensure,
+    // that it can't happen
+    if (isa<CoroAwaitSuspendInst>(I))
+      continue;
```
Should this be covered by the following check?

fpasserby replied:
No, because `CoroAwaitSuspendInst` is not a proper `IntrinsicInst`, because it can be invoked. I'm not sure what else could be done about it (aside from deriving `IntrinisicInst` from `CallBase` instead of `CallInst`). Also it seems like special handling for it should be added to `InlineCost.cpp` and this intrinisic should return true in `mayLowerToFunctionCall`. Callback metadata should probabply be attached to the intrinsic, but I wasn't able to find a way to do it.

ChuanqiXu9 commented on llvm/lib/Transforms/Coroutines/CoroSplit.cpp:
```diff
@@ -1553,6 +1625,9 @@ static bool hasCallsInBlocksBetween(BasicBlock *SaveBB, BasicBlock *ResDesBB) {
 }
 
 static bool hasCallsBetween(Instruction *Save, Instruction *ResumeOrDestroy) {
+  if (Save == ResumeOrDestroy)
+    return false;
```
How can this be true?

Oh, I see you use it in `isSimpleHelper`.

ChuanqiXu9 commented on llvm/lib/Transforms/Coroutines/CoroSplit.cpp:
```diff
@@ -1575,6 +1650,36 @@ static bool hasCallsBetween(Instruction *Save, Instruction *ResumeOrDestroy) {
+// 2. There are no calls between any of the returns and helper entry that could
+// resume or destroy it
+// FIXME: perform more sophisiticated analysis?
+static bool isSimpleHelper(CoroAwaitSuspendInst *AWS) {
```
Why do we have to handle the case? I don't feel it is necessary. It looks like a minor optimization which don't cover a lot of case but require more work to do.

fpasserby replied:
This method was added so that previously available optimization is not lost after this change. If you really feel this is too minor, I will remove it.

ChuanqiXu9 commented on clang/test/CodeGenCoroutines/pr59181.cpp:
```diff
@@ -48,13 +48,14 @@ void foo() {
   bar(true);
 }
 
+// FIXME: the test doesn't seem to be relevant anymore,
+// because objects that require cleanup are no more emitted in the suspend block
```
Testing is always good. Let's remove such comments.

ChuanqiXu9 commented on clang/test/CodeGenCoroutines/coro-symmetric-transfer-01.cpp:
```diff
@@ -47,13 +47,9 @@ detached_task foo() {
   co_return;
 }
 
+// FIXME: is this test needed now?
```
Let's remove the comment. Testing is always good.

ChuanqiXu9 reviewed (commented):
I got the reason why I felt `__await_suspend_helper_` was odd. In my imagination, we only need to emit the function `await_suspend(handle)` to an LLVM function and pass that to `llvm.coro.await.suspend` directly.

But the conversion to the actual code may need some additional code. So the idea to wrap them into a helper function makes sense. I feel the name `wrapper` sounds better to me. `Helper` is too broad.

ChuanqiXu9 commented on clang/lib/CodeGen/CGCoroutine.cpp:
```diff
@@ -338,6 +386,85 @@ static QualType getCoroutineSuspendExprReturnType(const ASTContext &Ctx,
+  Fn->setMustProgress();
+  Fn->addFnAttr(llvm::Attribute::AttrKind::AlwaysInline);
+
+  if (!CanThrow) {
+    Fn->addFnAttr(llvm::Attribute::AttrKind::NoUnwind);
+  }
```
We don't need this. If I remember correctly, LLVM can add it automatically as long as the await_suspend is marked noexcept.

ChuanqiXu9 commented on clang/lib/CodeGen/CGCoroutine.cpp:
```diff
@@ -232,16 +237,59 @@ static LValueOrRValue emitSuspendExpression(CodeGenFunction &CGF, CGCoroData &Co
+      IID = llvm::Intrinsic::coro_await_suspend_handle;
+
+    llvm::Function *AwaitSuspendIntrinsic = CGF.CGM.getIntrinsic(IID);
+    // FIXME: add call attributes?
+    if (AwaitSuspendCanThrow)
+      SuspendRet =
+          CGF.EmitCallOrInvoke(AwaitSuspendIntrinsic, SuspendHelperCallArgs);
+    else
+      SuspendRet = CGF.EmitNounwindRuntimeCall(AwaitSuspendIntrinsic,
+                                               SuspendHelperCallArgs);
```
CGF.EmitCallOrInvoke should be sufficient. We don't need to perform such optimizations here.

ChuanqiXu9 commented:
BTW, in my original comment, I said:

> Then in the middle end, we can perform analysis the awaitSuspendFn to see if the coroutine handle escapes or not. If it won't escape, we can replace above intrinsic to the direct call. Otherwise, we will only convert them in the CoroSplit phase.

But this patch only implements the otherwise part. This is not a blocking issue. But we need a TODO or a FIXME here.

fpasserby committed:
Get rid of OpaqueFramePtr

fpasserby committed:
Rename coro.await.suspend.void

fpasserby committed:
Switch on await_suspend return type

fpasserby committed:
Don't mark await_suspend_helper as nounwind

fpasserby committed:
Don't inline in CoroSplit

fpasserby committed:
Improve docs

fpasserby committed:
Rename to await.suspend.void in one more place

fpasserby committed:
Improve lowering

fpasserby committed:
Rename helper -> wrapper

fpasserby committed:
Remove unnecessary comments

fpasserby commented on llvm/docs/Coroutines.rst:
```diff
@@ -1744,6 +1744,273 @@ a call to ``llvm.coro.suspend.retcon`` after resuming abnormally.
 
+.. _coro.await.suspend:
+
+'llvm.coro.await.suspend' Intrinsic
```
Renamed to `coro.await.suspend.void`

fpasserby commented on clang/include/clang/AST/ExprCXX.h:
```diff
@@ -5036,14 +5036,17 @@ class CoroutineSuspendExpr : public Expr {
 
   Stmt *SubExprs[SubExpr::Count];
   OpaqueValueExpr *OpaqueValue = nullptr;
+  OpaqueValueExpr *OpaqueFramePtr = nullptr;
```
Now `__builtin_coro_frame` is used.

fpasserby committed:
Add await suspend lowering tests

fpasserby committed:
Remove unnecessary changes from SemaCoroutine

ChuanqiXu9 commented on clang/lib/Sema/SemaCoroutine.cpp:
```diff
@@ -380,66 +380,7 @@ static Expr *maybeTailCall(Sema &S, QualType RetType, Expr *E,
-  // looks relatively tolerant.
-  FD->addAttr(
-      NoInlineAttr::CreateImplicit(S.getASTContext(), FD->getLocation()));
+  return S.MaybeCreateExprWithCleanups(JustAddress);
```
We also need to update the comments of the function after this patch. Also I feel like the above fixme can be removed.

ChuanqiXu9 commented on llvm/lib/Transforms/Coroutines/CoroSplit.cpp:
```diff
@@ -167,6 +167,47 @@ class CoroCloner {
 } // end anonymous namespace
 
+// FIXME:
+// Lower the intrinisc earlier if coroutine frame doesn't escape
```
```suggestion
// Lower the intrinisc in CoroEarly phase if coroutine frame doesn't escape
```

ChuanqiXu9 commented on llvm/lib/Transforms/Coroutines/CoroSplit.cpp:
```diff
@@ -2013,6 +2104,10 @@ splitCoroutine(Function &F, SmallVectorImpl<Function *> &Clones,
   buildCoroutineFrame(F, Shape, MaterializableCallback);
   replaceFrameSizeAndAlignment(Shape);
 
+  IRBuilder<> Builder(M.getContext());
```
```suggestion
  IRBuilder<> Builder(F.getContext());
```

then we can get rid of `Module`.

ChuanqiXu9 commented on llvm/lib/Transforms/Coroutines/CoroSplit.cpp:
```diff
@@ -1396,9 +1478,18 @@ static bool simplifySuspendPoint(CoroSuspendInst *Suspend,
+  // Check that frame directly always refers to the current coroutine,
+  // either directly or via wrapper
+  if (Frame != CoroBegin) {
+    auto *AWS = dyn_cast<CoroAwaitSuspendInst>(Frame);
+    if (!AWS)
+      return false;
```
Since we've lowered all `CoroAwaitSuspendInst` before invoking `splitCoroutine`. We shouldn't see `CoroAwaitSuspendInst` here, right?

Then we can remove  isSimpleWrapper too.

ChuanqiXu9 reviewed (commented):
Thanks. This looks much better now.

Given the CoroAwaitSuspendInst is lowered before splitting coroutines, I think we don't need to handle it specially in `CoroSplit` any more.

fpasserby committed:
Review fixes

ChuanqiXu9 commented on clang/lib/CodeGen/CGCoroutine.cpp:
```diff
@@ -338,6 +414,71 @@ static QualType getCoroutineSuspendExprReturnType(const ASTContext &Ctx,
+  CurAwaitSuspendWrapper.FramePtr =
+      Builder.CreateLoad(GetAddrOfLocalVar(&FrameDecl));
+
+  // FIXME: mark as aliasing with frame?
+  // FIXME: TBAA?
+  // FIXME: emit in a better way (maybe egenerate AST in SemaCoroutine)?
```
Generally we like comments in a more formal style. Please rewrite the comments to make it more expressive.

ChuanqiXu9 reviewed (commented):
Thanks. This looks good to me except few nit comments.

Have you tested on a real world workloads?

fpasserby commented:
> Thanks. This looks good to me except few nit comments.
> 
> Have you tested on a real world workloads?

I've tested it on seastar and libcppcoro tests and on the example from #72006.

fpasserby committed:
Clarify comments

fpasserby committed:
Rename StmtCanThrow

ChuanqiXu9 reviewed (approved):
LGTM then. Thanks.

ChuanqiXu9 commented:
I'll merge this since I find you may not have commit access.

ahatanak commented on clang/lib/CodeGen/CGCoroutine.cpp:
```diff
@@ -338,6 +409,69 @@ static QualType getCoroutineSuspendExprReturnType(const ASTContext &Ctx,
+  // FIXME: add TBAA metadata to the loads
+  llvm::Value *AwaiterPtr = Builder.CreateLoad(GetAddrOfLocalVar(&AwaiterDecl));
+  auto AwaiterLValue =
+      MakeNaturalAlignAddrLValue(AwaiterPtr, AwaiterDecl.getType());
```
I wonder whether the type passed to `MakeNaturalAlignAddrLValue` is correct. `AwaiterDecl.getType()` is `void *`, but shouldn't `S.getOpaqueValue()` be passed?

I think this was causing a libc++ test to crash when ubsan was enabled (see https://github.com/llvm/llvm-project/pull/86898).

`S.getOpaqueValue()` and `AwaiterLValue` are passed to `OpaqueValueMappingData::bind`. clang  mis-compiles when the lvalue has a larger alignment than the type of `S.getOpaqueValue()`.

https://github.com/llvm/llvm-project/pull/79712/files#diff-da3676b91275038b801aabe1e10b6fdedfbe50b679eac152080c2a26fbd8ec7dR458

ChuanqiXu9 replied:
Yeah.. it looks like it would be better to pass the type of `S.getOpaqueValue()`. The ` void*` is simply a placeholder.

Would you like to handle this? I feel you are the better option since you can test it with your own patch. Otherwise I'll try to make it myself.

ahatanak replied:
I have a patch: https://github.com/llvm/llvm-project/pull/87134

mstorsjo commented:
FYI, as a headsup - I've bisected a regression down to this PR/commit. Before this change, all libcxx tests pass on Windows/armv7, but after this change, a bunch of coroutine tests fail on this target - see https://github.com/llvm/llvm-project/issues/100101.
